File: PersonaFlowReader/src/dataAccess/dataTypes/TextList.java
package dataAccess.dataTypes;

import dataAccess.FileReadWriteUtils;
import dataAccess.Library;
import dataAccess.LongCharRandomAccessFile;

import javax.naming.OperationNotSupportedException;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.LinkedList;

public class TextList {

    private static final String CHARACTER_NAME_COLOR = "\u001B[33m";
    private static final String RESET_COLOR = "\u001B[0m";

    private static final String START_SPECIAL = "(*";
    private static final String END_SPECIAL = "*)";

    private static final String PARAM_SEPARATOR = ",";

    private int size;
    //private final String[] textList;
    private final LinkedList<String> textList;
    private final LinkedList<Integer> refList;

    public TextList() {
        this.textList = new LinkedList<>();
        this.refList = new LinkedList<>();
    }

    public TextList(int size, LinkedList<String> textList, LinkedList<Integer> refList) {
        this.size = size;
        this.textList = textList;
        this.refList = refList;
    }

    public static TextList readEncodedTextList(RandomAccessFile file, int addr, boolean oneLine) throws IOException {
        int pointerBK = (int) file.getFilePointer();
        file.seek(addr);

        int size = FileReadWriteUtils.readInt(file, ByteOrder.LITTLE_ENDIAN);

        LinkedList<Integer> refList = new LinkedList<>();
        LinkedList<String> textList = new LinkedList<>();
        for (int i = 0; i < size; i++) {
            int textRef = FileReadWriteUtils.readInt(file, ByteOrder.LITTLE_ENDIAN);
            refList.add(textRef);
            textList.add(getText(file, textRef, oneLine));
        }
        file.seek(pointerBK);
        return new TextList(size, textList, refList);
    }

    private static String getText(RandomAccessFile file, int address, boolean oneLine) throws IOException {
        int pointerBK = (int) file.getFilePointer();
        StringBuilder toRet = new StringBuilder("\"");
        file.seek(address);

        StringBuilder optionsComment = null;

        short data = FileReadWriteUtils.readShort(file, ByteOrder.BIG_ENDIAN);
        boolean name_color_set = false;
        while (data != (short)0xff01) {
            // it's an instruction
            if (Library.getInstance().TEXT_INSTRUCTIONS.containsKey(data)) {
                Library.TextInstruction instr = Library.getInstance().TEXT_INSTRUCTIONS.get(data);
                switch(instr) {
                    case AWAITING_INPUT:
                    case PLAYER_FIRST_NAME:
                    case PLAYER_NICKNAME:
                    case PLAYER_LAST_NAME:
                    case COIN_NUMBER:
                        toRet.append(START_SPECIAL).append(instr.name()).append(END_SPECIAL);
                        break;
                    case LINE_BREAK:
                        if (name_color_set) {
                            if (!oneLine) toRet.append(RESET_COLOR);
                            name_color_set = false;
                        }
                        if (!oneLine) toRet.append("\n");
                        else toRet.append(START_SPECIAL).append(instr.name()).append(END_SPECIAL);
                        break;
                    case CONTINUE:
                        if (!oneLine) toRet.append("\n\n");
                        else toRet.append(START_SPECIAL).append(instr.name()).append(END_SPECIAL);
                        break;
                    case WAIT:
                    case PRINT_VALUE:
                        toRet.append(START_SPECIAL).append(instr.name()).append(PARAM_SEPARATOR);
                        data = FileReadWriteUtils.readShort(file, ByteOrder.LITTLE_ENDIAN);
                        toRet.append(data).append(END_SPECIAL);
                        break;
                    case SHOW_OPTIONS:
                        toRet.append(START_SPECIAL).append(instr.name()).append(PARAM_SEPARATOR);
                        data = FileReadWriteUtils.readShort(file, ByteOrder.LITTLE_ENDIAN);
                        toRet.append(data).append(END_SPECIAL);

                        if (optionsComment == null) {
                            optionsComment = new StringBuilder();
                        }
                        optionsComment.append("\t").append(Library.COMMENT_SYMBOL).append(" Shows options: |");
                        String[] options = Library.OPTIONS[data];
                        for (String s : options) {
                            optionsComment.append("\"").append(s).append("\"");
                            optionsComment.append("|");
                        }

                        break;
                    case SET_COLOR: // sets a color
                    case LEGACY_SET_COLOR:
                        data = FileReadWriteUtils.readShort(file, ByteOrder.LITTLE_ENDIAN);
                        if (!oneLine) {
                            toRet.append(Library.getInstance().TEXT_COLORS_ANSI.get(data));
                        }
                        else {
                            toRet.append(START_SPECIAL)
                                    .append(instr.name()).append(PARAM_SEPARATOR).append(Library.getInstance().TEXT_COLORS.get(data))
                                    .append(END_SPECIAL);
                        }
                        break;
                    case PRINT_ICON:
                        data = FileReadWriteUtils.readShort(file, ByteOrder.LITTLE_ENDIAN);
                        if (!oneLine) {
                            switch(data) {
                                case 0:
                                    toRet.append("\uD83E\uDDF4");
                                    break;
                                case 3:
                                    toRet.append("\uD83D\uDD11");
                                    break;
                                default:
                                    toRet.append("❓");
                            }
                        } else {
                            toRet.append(START_SPECIAL)
                                    .append(instr.name()).append(PARAM_SEPARATOR).append(data)
                                    .append(END_SPECIAL);
                        }
                        break;

                    case CHARACTER_NAME: // yellow
                        if (!oneLine) toRet.append(CHARACTER_NAME_COLOR);
                        else toRet.append(START_SPECIAL).append(instr.name()).append(END_SPECIAL);

                        name_color_set = true;
                        break;
                }
            } else { // it's text
                if (Library.getInstance().TEXT_CODES.containsKey(data)) {
                    toRet.append(Library.getInstance().TEXT_CODES.get(data));
                } else {
                    toRet.append(String.format("{%04X}", data));
                }
            }
            data = FileReadWriteUtils.readShort(file, ByteOrder.BIG_ENDIAN);
        }

        toRet.append("\"");

        if (optionsComment != null) {
            toRet.append(optionsComment);
        }

        // restore pointer position
        file.seek(pointerBK);
        return toRet.toString();
    }

    public static void encodeText(RandomAccessFile outputFile, String text, RandomAccessFile inputFile, int textInputPointer) throws IOException, OperationNotSupportedException {
        long pointerBK = inputFile.getFilePointer();

        //int length = text.length();
        int length = text.length();
        boolean canBeSpecial = false;
        boolean isSpecial = false;
        boolean canBeClosing = false;

        short data;
        StringBuilder special = new StringBuilder();
        for (int i = 0; i < length; i++) {
            String currChar = text.substring(i, i+1);
            //byte currChar = text[i];
            if (currChar.charAt(0) == START_SPECIAL.charAt(0)) {
                canBeSpecial = true;
                continue;
            } else if (canBeSpecial && currChar.charAt(0) == START_SPECIAL.charAt(1)) {
                isSpecial = true;
                canBeSpecial = false;
                continue;
                // if it ended up not being a special char
            } else if (canBeSpecial && !isSpecial) {
                // write the old chars
                data = Library.getInstance().TEXT_CODES_REVERSE.get("" + text.charAt(i-1));
                FileReadWriteUtils.writeShort(outputFile, ByteOrder.BIG_ENDIAN, data);

                // decrement i to force the loop to read the current char
                i--;
                canBeSpecial = false;
                continue;
            } else if (isSpecial) {
                if (currChar.charAt(0) == END_SPECIAL.charAt(0)) {
                    canBeClosing = true;
                    continue;
                } else if (canBeClosing && currChar.charAt(0) == END_SPECIAL.charAt(1)) {
                    // if a normal instruction
                    if (Library.getInstance().TEXT_INSTRUCTIONS_REVERSE.containsKey(special.toString())) {
                        data = Library.getInstance().TEXT_INSTRUCTIONS_REVERSE.get(special.toString());
                        FileReadWriteUtils.writeShort(outputFile, ByteOrder.BIG_ENDIAN, data);
                    } else {
                        // it is either the WAIT or SHOW_OPTIONS or SET_COLOR or PRINT_ICON.... instructions, which has a parameter
                        String[] split = special.toString().split(PARAM_SEPARATOR);
                        data = Library.getInstance().TEXT_INSTRUCTIONS_REVERSE.get(split[0]);
                        FileReadWriteUtils.writeShort(outputFile, ByteOrder.BIG_ENDIAN, data);
                        if (split[0].compareTo(Library.TextInstruction.WAIT.name()) == 0 ||
                                split[0].compareTo(Library.TextInstruction.SHOW_OPTIONS.name()) == 0 ||
                                split[0].compareTo(Library.TextInstruction.PRINT_VALUE.name()) == 0 ||
                                split[0].compareTo(Library.TextInstruction.PRINT_ICON.name()) == 0) {
                            data = Short.parseShort(split[1]); // the value
                            FileReadWriteUtils.writeShort(outputFile, ByteOrder.LITTLE_ENDIAN, data);
                        } else if (split[0].compareTo(Library.TextInstruction.SET_COLOR.name()) == 0 ||
                                split[0].compareTo(Library.TextInstruction.LEGACY_SET_COLOR.name()) == 0) {
                            data = Library.getInstance().TEXT_COLORS_REVERSE.get(split[1].charAt(0)); // the color
                            FileReadWriteUtils.writeShort(outputFile, ByteOrder.LITTLE_ENDIAN, data);
                        } else {
                            System.out.println("EERRROOOOORRRR in special text code\n");
                        }
                    }
                    special = new StringBuilder();
                    canBeSpecial = false;
                    isSpecial = false;
                    canBeClosing = false;
                    continue;
                }
                special.append(currChar);
                continue;
            }

            // When RandomAccessFile reads more complex characters, it puts them
            if ((byte)currChar.charAt(0) < 0) {
                LongCharRandomAccessFile longCharFile = new LongCharRandomAccessFile(textInputPointer+i, inputFile);
                inputFile.seek(textInputPointer+i);
                currChar = "" + longCharFile.nextChar();
                i += longCharFile.numOfBytes-1;
            }
            if (!Library.getInstance().TEXT_CODES_REVERSE.containsKey(currChar)) {
                currChar = convertCharToUsable(currChar);
            }
            data = Library.getInstance().TEXT_CODES_REVERSE.get(currChar);

            // TODO: remove this if when accents are figured out
            if (data > 0x006C && data <0x0078 ||
            data > 0x007E && data < 0x0088 ||
            data > 0x08E && data < 0x009A ||
            data > 0x00A8 && data < 0x00B0 ||
            data > 0x00BC && data < 0x00CF) {
                currChar = convertCharToUsable(currChar);
                data = Library.getInstance().TEXT_CODES_REVERSE.get(currChar);
            }
            FileReadWriteUtils.writeShort(outputFile, ByteOrder.BIG_ENDIAN, data);

        }
        FileReadWriteUtils.writeShort(outputFile, ByteOrder.BIG_ENDIAN, Library.getInstance().TEXT_INSTRUCTIONS_REVERSE.get("END"));
        inputFile.seek(pointerBK);
    }

    private static String convertCharToUsable(String currChar) throws OperationNotSupportedException {
        switch (currChar) {
            case "'":
                return "’"; // this char is used for Apostrophe instead of the standard '
            case "\"":
                return "”"; // this char can be used for quotes instead of the standard "
            case "~":
                return "～"; // this char can be used for tilde instead of the standard ~
            case "〜": // full width version of ~
                return "～";
            case "*":
                return "＊"; // this char can be used for asterisk instead of the standard *
            case "！": // full width version of !
                return "!";
            case "？": // full width version of ?
                return "?";
            case "＞":
                return ">";
            case "ａ":
                return "a";
            case "ｂ":
                return "b";
            case "ｃ":
                return "c";
            case "ｄ":
                return "d";
            case "ｅ":
                return "e";
            case "ｆ":
                return "f";
            case "ｇ":
                return "g";
            case "ｈ":
                return "h";
            case "ｉ":
                return "i";
            case "ｊ":
                return "j";
            case "ｋ":
                return "k";
            case "ｌ":
                return "l";
            case "ｍ":
                return "m";
            case "ｎ":
                return "n";
            case "ｏ":
                return "o";
            case "ｐ":
                return "p";
            case "ｑ":
                return "q";
            case "ｒ":
                return "r";
            case "ｓ":
                return "s";
            case "ｔ":
                return "t";
            case "ｕ":
                return "u";
            case "ｖ":
                return "v";
            case "ｗ":
                return "w";
            case "ｘ":
                return "x";
            case "ｙ":
                return "y";
            case "ｚ":
                return "z";

            case "Ａ":
                return "A";
            case "Ｂ":
                return "B";
            case "Ｃ":
                return "C";
            case "Ｄ":
                return "D";
            case "Ｅ":
                return "E";
            case "Ｆ":
                return "F";
            case "Ｇ":
                return "G";
            case "Ｈ":
                return "H";
            case "Ｉ":
                return "I";
            case "Ｊ":
                return "J";
            case "Ｋ":
                return "K";
            case "Ｌ":
                return "L";
            case "Ｍ":
                return "M";
            case "Ｎ":
                return "N";
            case "Ｏ":
                return "O";
            case "Ｐ":
                return "P";
            case "Ｑ":
                return "Q";
            case "Ｒ":
                return "R";
            case "Ｓ":
                return "S";
            case "Ｔ":
                return "T";
            case "Ｕ":
                return "U";
            case "Ｖ":
                return "V";
            case "Ｗ":
                return "W";
            case "Ｘ":
                return "X";
            case "Ｙ":
                return "Y";
            case "Ｚ":
                return "Z";

                // TODO: Delete all cases below when accents are figured out
            case "Á":
                return "A";
            case "À":
                return "A";
            case "Ã":
                return "A";
            case "Â":
                return "A";
            case "Ä":
                return "A";
            case "É":
                return "E";
            case "È":
                return "E";
            case "Ê":
                return "E";
            case "Ë":
                return "E";
            case "Í":
                return "I";
            case "Ì":
                return "I";
            case "Î":
                return "I";
            case "Ï":
                return "I";
            case "Ó":
                return "O";
            case "Ò":
                return "O";
            case "Õ":
                return "O";
            case "Ô":
                return "O";
            case "Ö":
                return "O";
            case "Ú":
                return "U";
            case "Ù":
                return "U";
            case "Û":
                return "U";
            case "Ü":
                return "U";
            case "Ç":
                return "C";
            case "á":
                return "a";
            case "à":
                return "a";
            case "ã":
                return "a";
            case "â":
                return "a";
            case "ä":
                return "a";
            case "é":
                return "e";
            case "è":
                return "e";
            case "ê":
                return "e";
            case "ë":
                return "e";
            case "ç":
                return "c";
            case "í":
                return "i";
            case "ì":
                return "i";
            case "î":
                return "i";
            case "ï":
                return "i";
            case "ó":
                return "o";
            case "ò":
                return "o";
            case "õ":
                return "o";
            case "ô":
                return "o";
            case "ö":
                return "o";
            case "ú":
                return "u";
            case "ù":
                return "u";
            case "û":
                return "u";
            case "ü":
                return "u";

            default:
                throw new OperationNotSupportedException("Character " + currChar + " is not usable.");
        }
    }

    @Override
    public String toString() {
        StringBuilder toRet = new StringBuilder();
        toRet.append("Number of dialogs: ").append(this.size).append("\n\n");
        for (int i = 0; i < this.size; i++) {
            //String text = new String(this.textList[i], StandardCharsets.UTF_8);
            String text = this.textList.get(i);
            toRet.append(i).append("(").append(String.format("0x%08x", refList.get(i))).append(")").append(":\n")
                    .append(text).append("\n\n");
        }
        return toRet.toString();
    }

    public void writeText(RandomAccessFile outputFile) throws IOException {
        //toRet.append("section\t.text\n");
        outputFile.writeBytes(Library.SECTION_KEYWORD + "\t" + Library.TEXT_AREA_KEYWORD + "\t" + String.format("0x%08x", this.size) + "\n");
        for (int i = 0; i < this.size; i++) {
            //toRet.append("\t").append(String.format("%03d", i)).append("\"").append((this.textList[i])).append("\"\n");
            outputFile.writeBytes(String.format("\t%03d:", i));
            outputFile.write(textList.get(i).getBytes());
            outputFile.writeBytes("\n");
        }
        //return toRet.toString();
    }

    public int indexOfText(int address) {
        for (int i = 0; i < size; i++) {
            if (this.refList.get(i) == address) {
                return i;
            }
        }
        return -1;
    }

    public void addText(RandomAccessFile file, int address, boolean oneLine) throws IOException {
        String text = getText(file, address, oneLine);
        this.refList.add(address);
        this.textList.add(text);

        this.size++;
    }

    public int getFirstAddr() {
        return this.refList.get(0);
    }

    public int getRef(int idx) {
        return this.refList.get(idx);
    }
}

File: PersonaFlowReader/src/dataAccess/dataTypes/TextList.kt
package dataAccess.dataTypes

import dataAccess.*
import java.io.IOException
import java.io.RandomAccessFile
import java.nio.ByteOrder
import java.text.Normalizer
import java.util.*
import javax.naming.OperationNotSupportedException

private const val CHARACTER_NAME_COLOR = "\u001B[33m"
private const val RESET_COLOR = "\u001B[0m"

private const val START_SPECIAL = "(*"
private const val END_SPECIAL = "*)"

private const val PARAM_SEPARATOR = ","

data class TextList(
    val textList: MutableList<String> = emptyList<String>().toMutableList(),
    val refList: MutableList<Int> = emptyList<Int>().toMutableList(),
    var size: Int = 0
) {
    override fun toString(): String {
        val toRet = java.lang.StringBuilder()
        toRet.append("Number of dialogs: ").append(this.size).append("\n\n")
        for (i in 0..<this.size) {
            //String text = new String(this.textList[i], StandardCharsets.UTF_8);
            val text = textList[i]
            toRet.append(i).append("(").append(String.format("0x%08x", refList[i])).append(")").append(":\n")
                .append(text).append("\n\n")
        }
        return toRet.toString()
    }

    @Throws(IOException::class)
    fun writeText(outputFile: RandomAccessFile) {
        //outputFile.writeBytes(Library.SECTION_KEYWORD + "\t" + Library.TEXT_AREA_KEYWORD + "\t" + String.format("0x%08x", this.size) + "\n");
        for (i in 0..<this.size) {
            //outputFile.writeBytes(String.format("\t%03d:", i));
            outputFile.write(textList[i].toByteArray())
            outputFile.writeBytes("\n")
        }
        //return toRet.toString();
    }

    fun indexOfText(address: Int): Int {
        return refList.indexOf(address)
        /*
        for (i in 0..<size) {
            if (refList[i] == address) {
                return i
            }
        }
        return -1
        */
    }

    @Throws(IOException::class)
    fun addText(file: RandomAccessFile, address: Int, oneLine: Boolean) {
        val text: String = getText(file, address, oneLine)
        refList.add(address)
        textList.add(text)

        size++
    }

    fun getFirstAddr(): Int {
        return refList[0]
    }

    fun getRef(idx: Int): Int {
        return refList[idx]
    }
}

fun readEncodedTextList(file: RandomAccessFile, addr: Int, oneLine: Boolean): TextList {
    val pointerBK = file.filePointer.toInt()
    file.seek(addr.toLong())

    val size = readInt(file, ByteOrder.LITTLE_ENDIAN)

    val refList = LinkedList<Int>()
    val textList = LinkedList<String>()
    for (i in 0..<size) {
        val textRef = readInt(file, ByteOrder.LITTLE_ENDIAN)
        refList.add(textRef)
        textList.add(getText(file, textRef, oneLine))
    }
    file.seek(pointerBK.toLong())
    return TextList(textList, refList, size)
}

@Throws(IOException::class)
private fun getText(file: RandomAccessFile, address: Int, oneLine: Boolean): String {
    val pointerBK = file.filePointer.toInt()
    val toRet = StringBuilder("\"")
    file.seek(address.toLong())

    var optionsComment: StringBuilder? = null

    var data = readShort(file, ByteOrder.BIG_ENDIAN)
    var name_color_set = false
    while (data != 0xff01.toShort()) {
        getLibInstance()
        // it's an instruction
        if (TEXT_INSTRUCTIONS.containsKey(data)) {
            val instr = TEXT_INSTRUCTIONS[data]
            when (instr) {
                TextInstruction.AWAITING_INPUT,
                TextInstruction.PLAYER_FIRST_NAME,
                TextInstruction.PLAYER_NICKNAME,
                TextInstruction.PLAYER_LAST_NAME,
                TextInstruction.COIN_NUMBER -> toRet.append(
                    START_SPECIAL
                ).append(instr.name).append(END_SPECIAL)

                TextInstruction.LINE_BREAK -> {
                    if (name_color_set) {
                        if (!oneLine) toRet.append(RESET_COLOR)
                        name_color_set = false
                    }
                    if (!oneLine) toRet.append("\n")
                    else toRet.append(START_SPECIAL).append(instr.name).append(END_SPECIAL)
                }

                TextInstruction.CONTINUE -> if (!oneLine) toRet.append("\n\n")
                else toRet.append(START_SPECIAL).append(instr.name).append(END_SPECIAL)

                TextInstruction.WAIT,
                TextInstruction.PRINT_VALUE -> {
                    toRet.append(START_SPECIAL).append(instr.name).append(PARAM_SEPARATOR)
                    data = readShort(file, ByteOrder.LITTLE_ENDIAN)
                    toRet.append(data.toInt()).append(END_SPECIAL)
                }

                TextInstruction.SHOW_OPTIONS -> {
                    toRet.append(START_SPECIAL).append(instr.name).append(PARAM_SEPARATOR)
                    data = readShort(file, ByteOrder.LITTLE_ENDIAN)
                    toRet.append(data.toInt()).append(END_SPECIAL)

                    if (optionsComment == null) {
                        optionsComment = StringBuilder()
                    }
                    optionsComment.append("\t").append(COMMENT_SYMBOL).append(" Shows options: |")
                    val options = OPTIONS[data.toInt()]
                    for (s in options) {
                        optionsComment.append("\"").append(s).append("\"")
                        optionsComment.append("|")
                    }
                }

                TextInstruction.SET_COLOR,
                TextInstruction.LEGACY_SET_COLOR -> {
                    data = readShort(file, ByteOrder.LITTLE_ENDIAN)
                    if (!oneLine) {
                        toRet.append(TEXT_COLORS_ANSI[data])
                    } else {
                        toRet.append(START_SPECIAL)
                            .append(instr.name).append(PARAM_SEPARATOR)
                            .append(TEXT_COLORS[data])
                            .append(END_SPECIAL)
                    }
                }

                TextInstruction.PRINT_ICON -> {
                    data = readShort(file, ByteOrder.LITTLE_ENDIAN)
                    if (!oneLine) {
                        when (data.toInt()) {
                            0 -> toRet.append("\uD83E\uDDF4")
                            3 -> toRet.append("\uD83D\uDD11")
                            else -> toRet.append("❓")
                        }
                    } else {
                        toRet.append(START_SPECIAL)
                            .append(instr.name).append(PARAM_SEPARATOR).append(data.toInt())
                            .append(END_SPECIAL)
                    }
                }

                TextInstruction.CHARACTER_NAME -> {
                    if (!oneLine) toRet.append(CHARACTER_NAME_COLOR)
                    else toRet.append(START_SPECIAL).append(instr.name).append(END_SPECIAL)

                    name_color_set = true
                }

                TextInstruction.END -> println("END found... code shouldn't have reached this place")
                null -> println("NULL in TextList.getText")
            }
        } else { // it's text
            if (TEXT_CODES.containsKey(data)) {
                toRet.append(TEXT_CODES[data])
            } else {
                toRet.append(String.format("{%04X}", data))
            }
        }
        data = readShort(file, ByteOrder.BIG_ENDIAN)
    }

    toRet.append("\"")

    if (optionsComment != null) {
        toRet.append(optionsComment)
    }

    // restore pointer position
    file.seek(pointerBK.toLong())
    return toRet.toString()
}

@Throws(IOException::class, OperationNotSupportedException::class)
fun encodeText(outputFile: RandomAccessFile?, text: String, inputFile: RandomAccessFile, textInputPointer: Int) {
    val pointerBK = inputFile.filePointer

    //int length = text.length();
    val length = text.length
    var canBeSpecial = false
    var isSpecial = false
    var canBeClosing = false

    var data: Short
    var special = java.lang.StringBuilder()
    var i = 0
    while (i < length) {
        var currChar = text.substring(i, i + 1)
        //byte currChar = text[i];
        if (currChar[0] == START_SPECIAL[0]) {
            canBeSpecial = true
            i++
            continue
        } else if (canBeSpecial && currChar[0] == START_SPECIAL[1]) {
            isSpecial = true
            canBeSpecial = false
            i++
            continue
            // if it ended up not being a special char
        } else if (canBeSpecial && !isSpecial) {
            // write the old chars
            data = TEXT_CODES_REVERSE["" + text[i - 1]]!!
            writeShort(outputFile!!, ByteOrder.BIG_ENDIAN, data)

            // decrement i to force the loop to read the current char
            i--
            canBeSpecial = false
            i++
            continue
        } else if (isSpecial) {
            if (currChar[0] == END_SPECIAL[0]) {
                canBeClosing = true
                i++
                continue
            } else if (canBeClosing && currChar[0] == END_SPECIAL[1]) {
                // if a normal instruction
                if (TEXT_INSTRUCTIONS_REVERSE.containsKey(special.toString())) {
                    data = TEXT_INSTRUCTIONS_REVERSE[special.toString()]!!
                    writeShort(outputFile!!, ByteOrder.BIG_ENDIAN, data)
                } else {
                    // it is either the WAIT or SHOW_OPTIONS or SET_COLOR or PRINT_ICON.... instructions, which has a parameter
                    val split: List<String> = special.toString().split(PARAM_SEPARATOR)
                        //special.toString().split(PARAM_SEPARATOR.toRegex()).dropLastWhile { it.isEmpty() }
                    data = TEXT_INSTRUCTIONS_REVERSE[split[0]]!!
                    writeShort(outputFile!!, ByteOrder.BIG_ENDIAN, data)
                    if (split[0].compareTo(TextInstruction.WAIT.name) == 0 || split[0].compareTo(TextInstruction.SHOW_OPTIONS.name) == 0 || split[0].compareTo(
                            TextInstruction.PRINT_VALUE.name
                        ) == 0 || split[0].compareTo(TextInstruction.PRINT_ICON.name) == 0
                    ) {
                        data = split[1].toShort() // the value
                        writeShort(outputFile, ByteOrder.LITTLE_ENDIAN, data)
                    } else if (split[0].compareTo(TextInstruction.SET_COLOR.name) == 0 ||
                        split[0].compareTo(TextInstruction.LEGACY_SET_COLOR.name) == 0
                    ) {
                        data = TEXT_COLORS_REVERSE[split[1][0]]!! // the color
                        writeShort(outputFile, ByteOrder.LITTLE_ENDIAN, data)
                    } else {
                        println("EERRROOOOORRRR in special text code\n")
                    }
                }
                special = java.lang.StringBuilder()
                canBeSpecial = false
                isSpecial = false
                canBeClosing = false
                i++
                continue
            }
            special.append(currChar)
            i++
            continue
        }

        // When RandomAccessFile reads more complex characters, it puts them
        if (currChar[0].code.toByte() < 0) {
            val longCharFile = LongCharRandomAccessFile(textInputPointer + i, inputFile)
            inputFile.seek((textInputPointer + i).toLong())
            currChar = "" + longCharFile.nextChar()
            i += longCharFile.numOfBytes - 1
        }
        if (!TEXT_CODES_REVERSE.containsKey(currChar)) {
            currChar = convertCharToUsable(currChar)
        }
        data = TEXT_CODES_REVERSE[currChar]!!

        // TODO: remove this if when accents are figured out
        /*if (data > 0x006C && data <0x0078 ||
            data > 0x007E && data < 0x0088 ||
            data > 0x08E && data < 0x009A ||
            data > 0x00A8 && data < 0x00B0 ||
            data > 0x00BC && data < 0x00CF) {
                currChar = convertCharToUsable(currChar);
                data = Library.getInstance().TEXT_CODES_REVERSE.get(currChar);
            }*/
        writeShort(outputFile!!, ByteOrder.BIG_ENDIAN, data)


        i++
    }
    writeShort(outputFile!!, ByteOrder.BIG_ENDIAN, TEXT_INSTRUCTIONS_REVERSE["END"]!!)
    inputFile.seek(pointerBK)
}

@Throws(OperationNotSupportedException::class)
private fun convertCharToUsable(currChar: String): String {
    var normalizedChar = Normalizer.normalize(currChar, Normalizer.Form.NFKC)
    return if (TEXT_CODES_REVERSE.containsKey(normalizedChar)) {
        normalizedChar
    } else {
        throw OperationNotSupportedException("Character $currChar is not usable.")
    }
}